test/Microsoft.AspNetCore.Mvc.Razor.Test/TagHelpers/TagHelperComponentTagHelperTest.cs

```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Logging.Testing;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.TagHelpers
+{
+    public class TagHelperComponentTagHelperTest
+    {
+        [Fact]
+        public void Initialized_InOrder()
+        {
+            // Arrange            
+            var tagHelperContext = new TagHelperContext(
+                "head",
+                allAttributes: new TagHelperAttributeList(
+                    Enumerable.Empty<TagHelperAttribute>()),
+                items: new Dictionary<object, object>(),
+                uniqueId: "test");
+
+            var incrementer = 0;
+            var testTagHelperComponentTagHelper = new TestTagHelperComponentTagHelper(new List<ITagHelperComponent>()
+            {
+                new CallbackTagHelperComponent(
+                    order: 1,
+                    initCallback: () =>
+                    {
+                        Assert.Equal(0, incrementer);
+                        incrementer++;
+                    },
+                    processAsyncCallback: null),
+                new CallbackTagHelperComponent(
+                    order: 2,
+                    initCallback: () =>
+                    {
+                        Assert.Equal(1, incrementer);
+                        incrementer++;
+                    },
+                    processAsyncCallback: null),
+                new CallbackTagHelperComponent(
+                    order: 3,
+                    initCallback: () =>
+                    {
+                        Assert.Equal(2, incrementer);
+                        incrementer++;
+                    },
+                    processAsyncCallback: null),
+            }, NullLoggerFactory.Instance);
+
+            // Act
+            testTagHelperComponentTagHelper.Init(tagHelperContext);
+
+            // Assert
+            Assert.Equal(3, incrementer);
+        }
+
+        [Fact]
+        public async void Processed_InOrder()
```

Comment:
`ProcessAsync_InvokesComponentsProcessAsyncInCorrectOrder`. Do the same for other classes below. 

To use this test case as an example; the value in being extra descriptive/explicit here is that we know that this method is testing the `ProcessAsync` of `TagHelperComponentTagHelper`. We also then know the test case is concerned with the underlying `ITagHelperComponents` ordering. If this test fails in the future the investigator will have all of that context to go on without having to deep dive into the test to determine what's wrong.